### PR TITLE
support colorTempPhysicalMin/Max

### PR DESCRIPTION
**Pull request by sjorge**
This PR is to implement #1928 but it turned out to have rather a lot of scope creep.

Opening this for some feedback so far it seems to work!

- configure() is used to read colorTempPhysicaMin/Max and colorCapabilities* while were at it 
- setting color_temp for a device now ensure to be withing the `ct_min-ct_max` range
- setting color_temp for a group is now ensured to be withing the `max(ct_min)-min(ct_max)` for all group members

*collecting these so we can try make meta.enhancedHue property obsolete in a follow up.

Note: I need  retesting as I have this on top of #2006 which I did first and tested separately.

### CODE REVIEW

**Comment by sjorge** [comment]
@Koenkk I guess we need to apply this too in scene_add and color_temp_move? 

Did I forget any other spots where we need to do this too... maybe the check/limiting should be move to a re-usable function? If so, whats a good spot for that? Also in lib/light I would think?

**Comment by sjorge** [comment]
Converted to a draft PR, as that is apparently an option!

File: converters/toZigbee.js [hunk 1]
```diff
@@ -566,11 +566,51 @@ const converters = {
     light_colortemp: {
         key: ['color_temp', 'color_temp_percent'],
         convertSet: async (entity, key, value, meta) => {
+            let colorTempMin;
+            let colorTempMax;
+            if (entity.constructor.name === 'Group' && entity.members.length > 0) {
```
Comment (Koenkk) [inline converters/toZigbee.js:571]: This can be simplified to something like this:

```js
const minCandidates = entity.members.map(m => m.getClusterAttributeValue('lightingColorCtrl', 'colorTempPhysicalMin')).filter(v => v !== undefined);
if (minCandidates.length === 0) {
   // log warning here
} else {
  colorTempMin = Math.min(...minCandidates);
}


```

**Comment by Koenkk** [comment]
> Did I forget any other spots where we need to do this too... maybe the check/limiting should be move to a re-usable function? If so, whats a good spot for that? Also in lib/light I would think?

Yes please do this, `lib/light.js` sounds good. Could be 1 function for block: https://github.com/Koenkk/zigbee-herdsman-converters/pull/2010/files#diff-a610ac853a94f2ed75dcf27fa22fb1cf9790749f9a0876e22c087a24e00885c4R569 and another one for https://github.com/Koenkk/zigbee-herdsman-converters/pull/2010/files#diff-a610ac853a94f2ed75dcf27fa22fb1cf9790749f9a0876e22c087a24e00885c4R604

File: converters/toZigbee.js [hunk 2]
```diff
@@ -566,11 +566,51 @@ const converters = {
     light_colortemp: {
         key: ['color_temp', 'color_temp_percent'],
         convertSet: async (entity, key, value, meta) => {
+            let colorTempMin;
```
Comment (Koenkk) [inline converters/toZigbee.js:569]: Extract this green block to a function in `lib/light.js`

File: converters/toZigbee.js [hunk 3]
```diff
@@ -566,11 +566,51 @@ const converters = {
...
+                    meta.logger.warn(`Missing colorTempPhysicalMin and/or colorTempPhysicalMax for ${entity.deviceIeeeAddress}!`);
+                }
+            }
+
             if (key === 'color_temp_percent') {
                 value = Number(value) * 3.46;
                 value = Math.round(value + 154).toString();
             }
 
+            // ensure value withing range
```
Comment (Koenkk) [inline converters/toZigbee.js:604]: Extract this green block to a function in `lib/light.js`

File: devices.js [hunk 4]
```diff
@@ -14771,7 +14789,14 @@ const devices = [
...
     if (device.extend) {
+        let deviceMeta = null;
+        if (device.extend.hasOwnProperty('meta') && device.hasOwnProperty('meta')) {
+            deviceMeta = Object.assign({}, device.extend.meta, device.meta);
+        }
         device = Object.assign({}, device.extend, device);
+        if (deviceMeta !== null) {
+            device.meta = deviceMeta;
+        }
         delete device.extend;
```
Comment (Koenkk) [inline devices.js:14800]: We should check if the configure is not overridden here. 


```js
if (device.extend.hasOwnProperty('configure') && device.hasOwnProperty('configure')) {
  assert.fail(`'${device.model}' has configure in extend and device, this is not allowed`);
}
```

If `npm tst` passes you are good!

Comment (sjorge) [reply devices.js:14800]: There is one device, I already had it on my too look at list.

https://github.com/Koenkk/zigbee-herdsman-converters/blob/de0abee0297f9fe95c1bfaa68562378ea8ad1f7a/devices.js#L10982

I'll give this another pass later this week.

File: converters/toZigbee.js (hunk 1 above)
Comment (sjorge) [reply converters/toZigbee.js:571]: Looks good, will try.

Also you made the same logical error I made in the beginning. We need `max` of the colorTempPhysicalMin and `min` of colorTempPhysicalMax for groups. To have the most restrictive range, unless I am wrong.

dev1: min=5, max=25
dev2: min=10, max=50

The range that is valid for everyone is 10-25 not 5-50.

Comment (Koenkk) [reply converters/toZigbee.js:571]: Yes, that's right indeed

File: devices.js (hunk 4 above)
Comment (sjorge) [reply devices.js:14800]: I tried adding the check, it fails... but then complains about assert not being defined.

```
[hyperon :: sjorge][~/Documents/repos/zigbee-herdsman-converters][ctminmax ✗]
[■]$ npm tst

> zigbee-herdsman-converters@14.0.1 test /Users/sjorge/Documents/repos/zigbee-herdsman-converters
> jest test

 PASS  test/otaCommon.test.js
 FAIL  test/index.test.js
  ● Test suite failed to run

    ReferenceError: assert is not defined

      14791 |     if (device.extend) {
      14792 |         if (device.extend.hasOwnProperty('configure') && device.hasOwnProperty('configure')) {
    > 14793 |             assert.fail(`'${device.model}' has configure in extend and device, this is not allowed`);
            |             ^
      14794 |         }
      14795 |
      14796 |         let deviceMeta = null;

      at map (devices.js:14793:13)
          at Array.map (<anonymous>)
      at Object.<anonymous> (devices.js:14790:26)
      at Object.<anonymous> (index.js:3:17)
      at Object.<anonymous> (test/index.test.js:1:1)

Test Suites: 1 failed, 1 passed, 2 total
Tests:       8 passed, 8 total
Snapshots:   0 total
Time:        2.11 s
Ran all test suites matching /test/i.
npm ERR! Test failed.  See above for more details.
[hyperon :: sjorge][~/Documents/repos/zigbee-herdsman-converters][ctminmax ✗]
[■]$ git diff
diff --git a/devices.js b/devices.js
index 8bce79e..10c273a 100755
--- a/devices.js
+++ b/devices.js
@@ -14789,14 +14789,21 @@ const devices = [

 module.exports = devices.map((device) => {
     if (device.extend) {
+        if (device.extend.hasOwnProperty('configure') && device.hasOwnProperty('configure')) {
+            assert.fail(`'${device.model}' has configure in extend and device, this is not allowed`);
+        }
+
         let deviceMeta = null;
         if (device.extend.hasOwnProperty('meta') && device.hasOwnProperty('meta')) {
             deviceMeta = Object.assign({}, device.extend.meta, device.meta);
         }
+
         device = Object.assign({}, device.extend, device);
+
         if (deviceMeta !== null) {
             device.meta = deviceMeta;
         }
+
         delete device.extend;
     }
```

Comment (sjorge) [reply devices.js:14800]: ```const assert = require('assert');``` might fix it, let me try again.

Comment (sjorge) [reply devices.js:14800]: @Koenkk I don't think the one with a conflict needs the configure on the device, I'll drop it but if you think it does need it. We can copy the fromZigbee/toZigbee converter list and add the read calls to the configure instead.

**Comment by sjorge** [comment]
So far, so good with the cleanup:

For a device:
```
Zigbee2MQTT:warn  2021-01-03 15:45:26: Requested color_temp 200 is lower than minimum supported 250, using minimum!
Zigbee2MQTT:debug 2021-01-03 15:45:26: Received Zigbee message from 'office/light/main/panel1', type 'attributeReport', cluster 'lightingColorCtrl', data '{"colorTemperature":250}' from endpoint 1 with groupID 0
```

```
Zigbee2MQTT:warn  2021-01-03 15:44:51: Requested color_temp 500 is higher than maximum supported 454, using maximum!
Zigbee2MQTT:info  2021-01-03 15:44:51: MQTT publish: topic 'zigbee2mqtt/office/light/main', payload '{"brightness":254,"color":{"x":0.458,"y":0.41},"color_temp":454,"state":"OFF"}'
```

For a group:
```
Zigbee2MQTT:warn  2021-01-03 15:46:53: Requested color_temp 200 is lower than minimum supported 250, using minimum!
Zigbee2MQTT:debug 2021-01-03 15:46:53: Received Zigbee message from 'office/light/main/panel4', type 'attributeReport', cluster 'lightingColorCtrl', data '{"colorTemperature":250}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2021-01-03 15:46:53: Received Zigbee message from 'office/light/main/panel3', type 'attributeReport', cluster 'lightingColorCtrl', data '{"colorTemperature":250}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2021-01-03 15:46:53: Received Zigbee message from 'office/light/main/panel1', type 'attributeReport', cluster 'lightingColorCtrl', data '{"colorTemperature":250}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2021-01-03 15:46:53: Received Zigbee message from 'office/light/main/panel2', type 'attributeReport', cluster 'lightingColorCtrl', data '{"colorTemperature":250}' from endpoint 1 with groupID 0
```

```
Zigbee2MQTT:warn  2021-01-03 15:46:55: Requested color_temp 500 is higher than maximum supported 454, using maximum!
Zigbee2MQTT:debug 2021-01-03 15:46:56: Received Zigbee message from 'office/light/main/panel2', type 'attributeReport', cluster 'lightingColorCtrl', data '{"colorTemperature":454}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2021-01-03 15:46:56: Received Zigbee message from 'office/light/main/panel4', type 'attributeReport', cluster 'lightingColorCtrl', data '{"colorTemperature":454}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2021-01-03 15:46:56: Received Zigbee message from 'office/light/main/panel1', type 'attributeReport', cluster 'lightingColorCtrl', data '{"colorTemperature":454}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2021-01-03 15:46:56: Received Zigbee message from 'office/light/main/panel3', type 'attributeReport', cluster 'lightingColorCtrl', data '{"colorTemperature":454}' from endpoint 1 with groupID 0
```

File: lib/light.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            (m) => m.getClusterAttributeValue('lightingColorCtrl', 'colorTempPhysicalMax'),
+        ).filter((v) => v !== undefined);
+        if (maxCandidates.length > 0) {
+            colorTempMax = Math.min(...maxCandidates);
+        }
+    } else {
+        colorTempMin = entity.getClusterAttributeValue('lightingColorCtrl', 'colorTempPhysicalMin');
+        colorTempMax = entity.getClusterAttributeValue('lightingColorCtrl', 'colorTempPhysicalMax');
+    }
+    if (logger !== undefined) {
```
Comment (Koenkk) [inline lib/light.js:29]: Logger is mandatory so this check can be removed.

File: lib/light.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+            const entityId = (entityType === 'group') ? entity.groupID : entity.deviceIeeeAddress;
+            logger.warn(`Missing colorTempPhysicalMin and/or colorTempPhysicalMax for ${entityType} ${entityId}!`);
+        }
+    }
+    return [colorTempMin, colorTempMax];
+}
+
+function clampColorTemp(colorTemp, colorTempMin, colorTempMax, logger) {
+    if ((colorTempMin !== undefined) && (colorTemp < colorTempMin)) {
+        if (logger !== undefined) {
```
Comment (Koenkk) [inline lib/light.js:41]: Same as https://github.com/Koenkk/zigbee-herdsman-converters/pull/2010/files#r551022328

File: lib/light.js [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+            logger.warn(`Requested color_temp ${colorTemp} is lower than minimum supported ${colorTempMin}, using minimum!`);
+        }
+        return colorTempMin;
+    }
+    if ((colorTempMax !== undefined) && (colorTemp > colorTempMax)) {
+        if (logger !== undefined) {
```
Comment (Koenkk) [inline lib/light.js:47]: Same as https://github.com/Koenkk/zigbee-herdsman-converters/pull/2010/files#r551022328

**Comment by Koenkk** [comment]
Besides 3 small comments, looks ready for merge to me! (if you are ready please undraft this pr).

**Comment by sjorge** [comment]
> Besides 3 small comments, looks ready for merge to me! (if you are ready please undraft this pr).

Done, I also updated scene_add.

```
Zigbee2MQTT:debug 2021-01-03 16:03:53: Received MQTT message on 'zigbee2mqtt/office/light/main/panel1/set' with data '{"scene_add": {"ID": 42, "state": "OFF", "color_temp": 100}}'
Zigbee2MQTT:debug 2021-01-03 16:03:53: Publishing 'set' 'scene_add' to 'office/light/main/panel1'
Zigbee2MQTT:warn  2021-01-03 16:03:53: Requested color_temp 100 is lower than minimum supported 250, using minimum!
```

**Comment by Koenkk** [comment]
Thanks 🎖️

**Comment by sjorge** [comment]
Yay! I'll try to have a look at auto detecting hue vs enhancedHue now we also have the colorCaps stored.
Hope to have a PR somewhere during the week... once I am done in my side quest wrt saturation.
